### DISCUSSION TITLE
Embedding templates in the Alembic repository

### DISCUSSION THREAD

**Discussion by matrixise** [sqlalchemy/alembic#1014] [Ideas]
Hi,

The idea of having a template for generating migration scripts is really great, but I think we can go further.

For example, embedding templates in a `./templates` directory that would contain several templates, one for table creation, or for adding columns. 

From there, we could specify which template we want to use when writing a new revision.

For example:

```
alembic revision --template add_table -m "Add User table"
alembic revision --template add_column_table -m "Add UUID for the user table"
```

I mention this because customizing a template script with all the options on the command line is becoming increasingly difficult to maintain.

**Comment by zzzeek** [comment]
this would be a straightforward feature to add and seems like it should be possible, though id like to understand your use case better.  what are "all the options on the command line" that you're passing?

**Reply by matrixise** [reply]
Hi @zzzeek 

Really sorry for the significant delay. I don't remember the reason for "all the options", but having a template could be a cool feature. the other point is if we could read some fields from the CLI (via input) or the CLI (dynamic options). 

I like the idea of CookieCutter where you have a .json file with the description of the fields for the template, and integration with Alembic could be cool.

Example
```
alembic revision --template add_custom_table -m "Add User table" --field:tablename=users
alembic revision --template add_uuid_column -m "Add UUID for the user table" --field:tablename=users
```

etc...

**Reply by CaselIT** [reply]
Personally I'm not sure how useful this would be.
I think most of the people who you use this would instead use autogenerate. There is also the issue that often a revision has multiple operations.

I think this could maybe make more sense as an extension library. On alembic side we could add additional API, if some are missing, to allow implementation of it

**Reply by matrixise** [reply]
> Personally I'm not sure how useful this would be. I think most of the people who you use this would instead use autogenerate. There is also the issue that often a revision has multiple operations.

Sure, a revision can have multiple operations, but in some cases, you can just want to create a new table, or add/drop/change the type of a field. The `script.py.mako` is just a small limitation.

Now, I think of another solution, create a shell script over `alembic` and this one could change a synlink, `script.py.mako` -> `templates/add_custom_table.py.mako`

> 
> I think this could maybe make more sense as an extension library. On alembic side we could add additional API, if some are missing, to allow implementation of it

+1

**Reply by zzzeek** [reply]
what's in these templates?  is the idea that you are not using autogenerate?

**Reply by matrixise** [reply]
You think we use DDL in the database and then use autogenerate, hoping that autogenerate supports the requested transformations. In some cases, it is not possible.

For example, renaming a table, a column, or changing the type of a varchar field to enum, or the other way around.

the dev team could use a list of templates in a README, or update the `script.py.mako` with commented blocks for patterns.

Or just ;-) use a template, `alembic --template convert_varchar_to_enum -m "Convert the status field of invoices to an enum"`

**Reply by zzzeek** [reply]
if the varchar to enum case is the main rationale, why dont we [fix that case](https://github.com/sqlalchemy/alembic/issues/278) ?  We likely need a dedicated alembic construct for that.

Having a "template" for varchar-> enum means you want to create a migration that does exactly one change to a single column.    if I were burdened by the "varchar->enum" case and I didnt have time to contribute to #278, I wouldn't want a template.  I'd make a Python function to do it!   The import for my python function would be part of my single alembic template so that I can use it when I need it.

Why "templates" and not "modules with helper functions" ?  the latter seems much more obvious and generally useful.

**Reply by zzzeek** [reply]
similarly, rename a table, why do we need a template for that?   [use the function](https://alembic.sqlalchemy.org/en/latest/ops.html?highlight=op+rename_table#alembic.operations.Operations.rename_table) ?  it's one line?  How would --template work with that ?  would there also be command line arguments for table name, schema name, column name, etc. ?   This seems way out of scope for alembic itself.

**Reply by zzzeek** [reply]
Here's screenshots of using vscode - rename table is right in the dropdown, argument names, works great, nothing new to learn outside of your IDE, this seems a lot easier to me if people are writing templates by hand...

![Screenshot from 2022-09-06 09-31-14](https://user-images.githubusercontent.com/128223/188648440-2037f41a-3348-4ee9-ba0e-f81437349fb5.png)
![Screenshot from 2022-09-06 09-33-29](https://user-images.githubusercontent.com/128223/188648645-ff1776eb-02ad-43b5-b2fc-946d93199ce4.png)

**Reply by matrixise** [reply]
it was a simple idea, I think we can close the issue.

**Reply by zzzeek** [reply]
whoops I hope I didnt offend, these are the questions I'd be asking of any feature that we add to alembic or SQLAlchemy.    As proposed the option seems simple and possibly useful but after many years of doing this I've learned that anything you add has to fit into a bigger story of how the library / application is going to be used.   otherwise you are just adding things for the sake of adding them, and possibly missing the best way for something to be added, which can then make it hard to change later.  so I was just looking for what these templates would be doing in terms of what kinds of problems you want to solve and how it would look in practice.
